mysite/templates/ct/lessons.html

```diff
@@ -136,28 +136,110 @@
 </div>
 {% endif %}
 
-{% if lessonSet %}
+{#{% if lessonSet %}#}
+{#<h2>Search Results</h2>#}
+{#If one of these is relevant to this concept,#}
+{#please click <b>{{ actionLabel }}</b>.#}
+{#<table class="table table-striped">#}
+{#<thead><tr>#}
+{#  <th>Lesson</th><th>Type</th><th>Author</th>#}
+{#</tr></thead>#}
+{#<tbody>#}
+{#{% for ul in lessonSet %}#}
+{#  <tr><td>#}
+{#  <a href="{{ actionTarget |get_object_url:ul }}">{{ ul.lesson.title }} {{ ul.treeID }} {{ ul.branch }}</a>#}
+{#  <form action="{{ actionTarget }}" method="post"#}
+{#   style=" display:inline!important;">#}
+{#  {% csrf_token %}#}
+{#  <input type="hidden" name="ulID" value="{{ ul.pk }}" />#}
+{#  <input type="submit" value="{{ actionLabel }}" />#}
+{#  </form>#}
+{#  </td>#}
+{#    <td>{{ ul.lesson.get_kind_display }}</td>#}
+{#    <td>{{ ul.lesson.addedBy.get_full_name }}</td>#}
+{#  </tr>#}
+{#{% endfor %}#}
+{#</tbody>#}
+{#</table>#}
+{#{% elif foundNothing %}#}
+{#  <b>No matches found</b>.  Try another search, or write a new lesson.#}
+{#{% endif %}#}
+
+{% if full_list %}
 <h2>Search Results</h2>
 If one of these is relevant to this concept,
 please click <b>{{ actionLabel }}</b>.
 <table class="table table-striped">
 <thead><tr>
-  <th>Lesson</th><th>Type</th><th>Author</th>
+  <th>Lesson</th><th></th><th>Type</th><th>Author</th>
 </tr></thead>
 <tbody>
-{% for ul in lessonSet %}
-  <tr><td>
-  <a href="{{ actionTarget |get_object_url:ul }}">{{ ul.lesson.title }}</a>
-  <form action="{{ actionTarget }}" method="post"
-   style=" display:inline!important;">
-  {% csrf_token %}
-  <input type="hidden" name="ulID" value="{{ ul.pk }}" />
-  <input type="submit" value="{{ actionLabel }}" />
-  </form>
-  </td>
-    <td>{{ ul.lesson.get_kind_display }}</td>
-    <td>{{ ul.lesson.addedBy.get_full_name }}</td>
-  </tr>
+{% for tree, branch in full_list.items %}
+    {% if branch %}
+        <tr><td colspan="4">
+        {{ tree.lesson.title }} &nbsp;
+    <a role="button" data-toggle="collapse" class="btn btn-default" href="#{{ tree.id }}" aria-expanded="false" aria-controls="{{ tree.id }}">
+        Versions
+         </a>
+        <hr>
+
+        <div class="collapse" id="{{ tree.id }}">
+            <table class="table table-striped">
+                <thead>
+                    <th>Lesson</th><th>Branch</th><th>Author</th><th>Commit message</th><th></th><th>Commit time</th><th>Type</th>
+                </thead>
+                <tr>
+                  <td>
+                    <a href="{{ actionTarget|get_object_url:tree }}">{{ tree.lesson.title }}</a>
+                  </td>
+                  <td>{{ tree.branch }}</td><td>{{ tree.lesson.addedBy.get_full_name }}</td><td>{{ tree.lesson.changeLog }}</td>
+                  <td>
+                  <form action="{{ actionTarget }}" method="post"
+                   style=" display:inline!important;" class="pull-right">
```

Comment:
> Can we avoid using Inline styles? Because, when templates grows, it will be hard to maintain all the inline definitions.

I agree - this is very bad. This will be changed on next refactoring step. Or these templates will be deprecated due to new UI plans (HackatonUI)